Adicionando 2975 (#202)
petbccufscar/URI: Strings/2975/README.md

# Problema:

Uma sequência de letras é escrita em um anel milagroso. Quando a sequência é lida em voz alta a partir de qualquer letra, a mágica acontece. Para obter a maior magia, você precisa da palavra mais poderosa: você quer encontrar um lugar para começar a ler, de modo que a palavra que você recebe seja a maior palavra possível na ordem alfabética.

**Problema Completo**: https://www.urionlinejudge.com.br/judge/pt/problems/view/2975

# Resolução:

Para a resolução deste problema, além da biblioteca padrão `stdio.h` (biblioteca para o scanf), é necessário incluir a biblioteca `string.h` (biblioteca para o strlen). Definimos uma constante para o máximo número de caracteres, e fazemos uma definição de macro para achar o mínimo entre dois elementos.

```c
#include <stdio.h>
#include <string.h>

#define MAX 100010

#define min(a, b) (a < b ? a : b)
```

> Abaixo, o temos o vetor maior prefixo que também é um sufixo, isto é, o comprimento de str[0...k] que é igual a [n-k...n-1].

Reviewer: Não entendi muito essa parte da explicação

> Abaixo, o temos o vetor maior prefixo que também é um sufixo, isto é, o comprimento de str[0...k] que é igual a [n-k...n-1].

Author: ```suggestion
Abaixo, o temos algoritmo LPS (longest palindromic subsequence) para retornar o comprimento da subsequência palíndrômica mais longa na sequência.
```

> Com len = 0 temos o comprimento atual, e lps[0] de zero é zero por definição.

Author: ```suggestion
Com len = 0 temos o comprimento atual, e o vetor lps na posição zero é zero por definição.
```

> 

Reviewer: lps[0] de zero ficou esquisito, acho melhor "o vetor lps na posição zero"

Em seguida, temos um loop para calcular o lps:

> Se for igual, incremento o comprimento e seto o lps[i];

Reviewer: igual a?

> Se for igual, incremento o comprimento e seto o lps[i];

Author: ```suggestion
Se pat[i] == pat[len] incremento o comprimento e seto o lps[i];
```

> Abaixo, o temos algoritmo LPS (longest palindromic subsequence) para retornar o comprimento da subsequência palíndrômica mais longa em sequência.
> 
> Com len = 0 temos o comprimento atual, e o vetor lps na posição zero é zero por definição.
> 
> Em seguida, temos um loop para calcular o lps:
> 
> Se pat[i] == pat[len] incremento o comprimento e seto o lps[i];
> 
> Se for diferente de zero, então volto para o lps anterior;
> 
> Se o comprimento for zero, não há o que fazer e o lps atual é nulo.

Reviewer: ```suggestion
Abaixo temos a função `computeLPSArray` que nos dá o comprimento da subsequência mais longa da sequência de caracteres `pat`.

Definimos `len` como o comprimento da subsequência buscada, já que inicialmente ela não existe. Também definimos 0 no ínicio do vetor `lps`.

```c
void computeLPSArray(char* pat, int M, int* lps) { 
    
    int len = 0; 
  
    lps[0] = 0;
    /*
    .
    .
    */
}
```

Agora, criamos um loop para calcular a longest palindromic subsequence, enquanto nosso contador for menor que `M` (tamanho de `pat`) fazemos ações de acordo com verificações.
* Caso `pat[i] == pat[len]`, fazemos X
* Caso em que são diferentes e o valor de `len` já foi alterado, ou seja `len != 0`, fazemos Y
* Caso em que são diferentes porém o valor de `len` ainda é 0, fazemos Z

Dessa forma, nossa função ao final consegue calcular ABCDEFGHI.

```c
while (i < M) { 
        if (pat[i] == pat[len]) { 
            len++; 
            lps[i] = len; 
            i++; 
        } 
        else {  
            if (len != 0) {
                len = lps[len - 1]; 
            } 
            else {
             
                lps[i] = 0; 
                i++; 
            } 
        } 
    } 
```

                lps[i] = 0; 
                i++; 
            } 
        } 
    } 
} 
```

Abaixo, temos o algoritmo KMP pra encontrar todas ocorrências de uma string em outra em tempo linear.

O algoritmo de Knuth–Morris–Pratt procura a ocorrência de uma "palavra" W dentro de um "texto" S empregando a simples técnica de que quando aparece uma diferença, a palavra tem em si a informação necessária para determinar onde começar a próxima comparação.

Segue a explicação do algoritmo, no qual inicialmente definimos os seguintes valores do tipo inteiro:

int ret = 0; quantidade de ocorrências / int M = strlen(pat); tamanho da string a ser pesquisada / int N = strlen(txt); tamanho da string local de pesquisa / int lps[M]; vetor lps da string padrão / int i = 0; E int j = 0; são contadores

E em seguida, após as definições tipo inteiro, temos o loop do KMP, que está explicado abaixo.

Temos o while (i < N-1) onde usamos -1 pois a string está duplicada devido ao ciclo.

Se for igual incremento os dois índices.

Se j for o tamanho da string padrão, ocorrência encontrada.

Com o j = lps[j - 1] vai para o índice maior prefixo que também é um sufixo, pois é invariante a pesquisa.

> Caso tenha dado errado, temos o else-if, no qual:

Reviewer: else if*

> Caso tenha dado errado, temos o else-if, no qual:

Author: ```suggestion
Caso tenha dado errado, temos o else if, no qual:
```

> Usamos o if (j != 0) para se o índice for j != 0, tento o lps[j-1].

Reviewer: para se o índice `j` for diferente de 0, tento o `lps[j-1]`

> Usamos o if (j != 0) para se o índice for j != 0, tento o lps[j-1].

Author: ```suggestion
Usamos o if (j != 0) para e o índice j for diferente de 0, tento o lps[j-1].
```

# Problema:

Uma sequência de letras é escrita em um anel milagroso. Quando a sequência é lida em voz alta a partir de qualquer letra, a mágica acontece. Para obter a maior magia, você precisa da palavra mais poderosa: você quer encontrar um lugar para começar a ler, de modo que a palavra que você recebe seja a maior palavra possível na ordem alfabética.

**Problema Completo**: https://www.urionlinejudge.com.br/judge/pt/problems/view/2975

# Resolução:

Para a resolução deste problema, além da biblioteca padrão `stdio.h` (biblioteca para o scanf), é necessário incluir a biblioteca `string.h` (biblioteca para o strlen). Definimos uma constante para o máximo número de caracteres, e fazemos uma definição de macro para achar o mínimo entre dois elementos.

```c
#include <stdio.h>
#include <string.h>

#define MAX 100010

#define min(a, b) (a < b ? a : b)
```

Abaixo temos a função `computeLPSArray` que nos dá o comprimento da subsequência mais longa da sequência de caracteres `pat`.

> Definimos `len` como o comprimento da subsequência buscada, já que inicialmente ela não existe. Também definimos 0 no ínicio do vetor `lps`.
> 
> ```c
> void computeLPSArray(char* pat, int M, int* lps) { 
>     
>     int len = 0; 
>   
>     lps[0] = 0;
>     /*
>     .
>     .
>     */
> }

Author: ```suggestion
Inicialmente definimos `len` como o comprimento da subsequência buscada, já que inicialmente ela não existe. Também definimos 0 no ínicio do vetor `lps`.

Em seguida criamos um loop para calcular a longest palindromic subsequence, enquanto nosso contador for menor que M (tamanho de `pat`) fazemos ações de acordo com verificações.

- Caso `pat[i] == pat[len]`, incremento o compriento e seto o `lps[i] = len`
- Caso em que são diferentes e o valor de `len` já foi alterado, ou seja `len != 0`, então volta para o lps anterior, ou seja `len = lps[len - 1]`.
- Caso em que são diferentes porém o valor de `len` ainda é 0, não há o que fazer e o lps atual é nulo, ou seja, `lps[i] = 0`.

Dessa forma, nossa função ao final consegue calcular ABCDEFGHI.
```
